### ISSUE TITLE
Chapel compilation only works once and it is not usable again

### ISSUE THREAD

**Issue by diehlpk** [chapel-lang/chapel#24644]
### Summary of Problem

I installed Chapel on an Arm-based machine and I used the interactive session to compile Chapel. I could compile Chapel and could run my examples. So good so far. However, closing the interactive session was not a very clever idea. Since my runs are long, I prepared batch file to submit. However, now I get the following error message

```
 cat error-185976.txt 
error: The runtime has not been built for this configuration. Run $CHPL_HOME/util/chplenv/printchplbuilds.py for information on available runtimes.
```

What behavior did you observe when encountering this issue?

I assumed once I compiled Chapel, I could use it in my batch job or another interactive session.

What behavior did you expect to observe?

To use my chapel installation multiple times. 

Is this a blocking issue with no known work-arounds?

Not I am ware off. One could write a python script to set the correct environment variables. However, that is not very user-friendy. 

### Steps to Reproduce

- Compile chapel
- Run chapel code in the same interactive session
- Close the interactive session
- Open a new interactive session or submit a batch job

### Configuration Information

- Output of `chpl --version`:
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:

**Comment by arezaii** [comment]
Thank you for reporting this issue, it sounds like you might need to setup your new environment (interactive session or batch job) to use the Chapel configuration you previously built. If you haven't already done that, please see https://chapel-lang.org/docs/usingchapel/building.html#using-chapel-in-a-different-shell-terminal-session and let us know if the suggestions there do not resolve the issue. Otherwise, please provide some more information, especially the output of `$CHPL_HOME/util/printchplenv --anonymize`.

**Comment by mppf** [comment]
> I installed Chapel on an Arm-based machine and I used the interactive session to compile Chapel. 

Did you use `./configure` or `make install` ? If you did, that should have created a `chplconfig` file that impacts the default configuration when you use the installed `chpl`. (And if that process didn't work somehow, it's a bug). If you didn't, that's surprising to me because I would have expected these commands to be run when installing.

The `chplconfig` file is one of the things mentioned in the documentation that @arezaii linked to above. (And there are other ways to address it as well). Note that, while `chplconfig` should be created as part of `./configure`, you can also create one separately if that is helpful to you.

**Comment by diehlpk** [comment]
@arezaii that is my output of `util/printchplenv  --anonymize`

```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: aarch64
CHPL_TARGET_CPU: unknown
CHPL_LOCALE_MODEL: flat *
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: bundled *
CHPL_MEM: jemalloc *
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: none
CHPL_HWLOC: bundled *
CHPL_RE2: bundled *
CHPL_LLVM: unset
CHPL_AUX_FILESYS: lustre *
```

**Comment by diehlpk** [comment]
Here comes the most confusing part to me, if I run `$CHPL_HOME/util/chplenv/printchplbuilds.py` I get 

```
                           <Current>              0                    1                 
     CHPL_TARGET_PLATFORM: linux64              linux64              linux64             
     CHPL_TARGET_COMPILER: llvm                 clang*               llvm                
         CHPL_TARGET_ARCH: aarch64              aarch64              aarch64             
          CHPL_TARGET_CPU: native               native               native              
        CHPL_LOCALE_MODEL: flat                 flat                 flat                
                CHPL_COMM: none                 none                 none                
               CHPL_TASKS: qthreads             fifo*                fifo*               
         CHPL_TASKS_DEBUG: -                    -                    -                   
              CHPL_TIMERS: generic              generic              generic             
              CHPL_UNWIND: none                 none                 none                
                 CHPL_MEM: jemalloc             cstdlib*             cstdlib*            
             CHPL_ATOMICS: cstdlib              cstdlib              cstdlib             
               CHPL_HWLOC: bundled              none*                none*               
         CHPL_HWLOC_DEBUG: -                    -                    -                   
                 CHPL_RE2: bundled              none*                none*               
         CHPL_AUX_FILESYS: none                 none                 none                
             CHPL_LIB_PIC: none                 none                 none                
        CHPL_SANITIZE_EXE: none                 none                 none                
                    MTIME: NA                   Mar 16 22:14         Mar 16 22:03 
```

**Comment by mppf** [comment]
@diehlpk -- based on the `printchplenv` you showed, your `chplconfig` should also include the variables printed with a `*` at the end (although some are unnecessary as they are the default). In particular, this is what I think your chplconfig file should contain:


```
CHPL_TARGET_CC=clang++
CHPL_LLVM=system
CHPL_LLVM_CONFIG=llvm-config
CHPL_COMM=gasnet
CHPL_UNWIND=bundled
CHPL_ATOMICS: cstdlib
CHPL_AUX_FILESYS=lustre
```

Are you able to update your `chplconfig` in this way? (if you did `make install` it ends up installed somewhere & you'd want to modify that one or re-install).

Assuming that works (and I would expect it to), the next question is -- why did the `./configure` process lead to the `chplconfig` that isn't what you wanted? One theory: maybe you ran `./configure` again with different environment variables set, in order to change the install prefix (or something). That would overwrite the `chplconfig` file. Another possibility is that you set environment variables like `CHPL_COMM` after you ran `./configure`. Or maybe it is something else. If we can figure out what happened, we can think about how to make it less likely to come up in the future.

**Comment by diehlpk** [comment]
@mppf Thanks, I edited my  file and edited `chplconfig` to 

```
CHPL_TARGET_CC=clang++
CHPL_LLVM=system
CHPL_LLVM_CONFIG=llvm-config
CHPL_COMM=gasnet
CHPL_UNWIND=bundled
CHPL_ATOMICS: cstdlib
CHPL_AUX_FILESYS=lustre
```

but now I get

```
warning: Unknown target CPU aarch64 -- not specializing
note: To see available CPU types, run clang++ --target=aarch64-unknown-linux-gnu --print-supported-cpus
error: The runtime has not been built for this configuration. Run $CHPL_HOME/util/chplenv/printchplbuilds.py for information on available runtimes.
```

Let me start a complete new build of chapel and delete my old one.

**Comment by diehlpk** [comment]
Ok, I cloned a fresh chapel GitHub repo and switched to 1.33.0, and did

```
module load llvm/16
export CC=clang
export CXX=clang++
./configure --prefix=/lustre/projects/diehl-group/chris/opt/chapel
```

and not I get the following configuration 

```
Compiling Python scripts in util/

  Currently selected Chapel configuration:

CHPL_TARGET_PLATFORM: aarch64 *
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: aarch64
CHPL_TARGET_CPU: unknown
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: bundled *
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_RE2: none
CHPL_LLVM: system *
CHPL_AUX_FILESYS: lustre *

  Selected installation options:

PREFIX: /lustre/projects/diehl-group/chris/opt/chapel
```

However, I can not compile hwloc, see below 

```
###
### Configuring hwloc core
###
checking hwloc building mode... standalone
configure: hwloc builddir: /lustre/projects/diehl-group/chris/chapel/third-party/hwloc/build/aarch64-aarch64-unknown-llvm-none-flat
configure: hwloc srcdir: /lustre/projects/diehl-group/chris/chapel/third-party/hwloc/hwloc-src
configure: Detected VPATH build
checking for hwloc version... 2.9.0
checking if want hwloc maintainer support... disabled
checking for hwloc directory prefix... (none)
checking for hwloc symbol prefix... chpl_
configure: WARNING: hwloc requires a C99 compiler
configure: error: Aborting.
make[4]: *** [Makefile:76: hwloc-config] Error 1
make[3]: *** [Makefile:123: /lustre/projects/diehl-group/chris/chapel/third-party/hwloc/install/aarch64-aarch64-unknown-llvm-none-flat] Error 2
make[2]: *** [Makefile:76: third-party-pkgs] Error 2
make[1]: *** [Makefile:100: runtime] Error 2
make: *** [Makefile:63: comprt] Error 2
```

Any idea what I did wrong?

**Comment by bradcray** [comment]
@diehlpk : I can never remember the interplay between the various CC/CXX variables offhand, and unfortunately @mppf has had to head out for the day.  Here's my attempt to pick up the conversation, where you might start with the last bullet below.

With respect to your latest error, could you share the output of `$CHPL_HOME/util/printchplenv --all --anonymize` (where I'm particularly interested in the `CHPL_TARGET_[CC|CXX|LD]` variables)?

Also, do you see an invocation of `./configure` for hwloc earlier in the log you excerpt here?

Then two other questions/observations:
* When you say you "cloned a fresh chapel GitHub repo and switched to 1.33.0", what do you mean by "switched to 1.33.0"?  Are you cloning a specific branch of the GitHub repo rather than the `main` branch?
* In your earlier comments, you mentioned `CHPL_TARGET_CC=clang++`, but I would've expected `CHPL_TARGET_CC=clang` or `CHPL_TARGET_CXX=clang++` not `CC` and `clang++` mixed together.  If you're still setting that in this latest run, that could be the core issue.

**Comment by diehlpk** [comment]
@bradcray I am doing this here 

`git clone --branch release/1.33 https://github.com/chapel-lang/chapel.git`

**Comment by bradcray** [comment]
@diehlpk : Also, if iterating more quickly on this over chat would be useful, and you have access to Gitter, we could pick this up here:  https://gitter.im/chapel-lang/chapel.  Or possibly in Slack if you want to mail me for a link.

**Comment by diehlpk** [comment]
@bradcray I joined Gitter and we can discuss there.

**Comment by bradcray** [comment]
Jumping back here, some of the issues we hit were:

* When `CHPL_TARGET_CC|CXX` were set to `clang|clang++` respectively, but `CHPL_TARGET_COMPILER` was not set to `clang` we were getting the unhelpful error `Error: Missing or wrong version for clang at clang`
* Running `printchplenv --all --anonymize` didn't work because it generated the same error as above, which made it difficult to determine what version of `CHPL_LLVM_CONFIG` the compiler was using
* Adding the setting `CHPL_TARGET_COMPILER=clang` generated `Error: Could not find clang with the same version as CHPL_LLVM_CONFIG=llvm-config-15`, where the version of clang/clang++ in the path was 16 (which was probably the cause of the error in the previous bullet).
* Unsetting all three of `CHPL_TARGET_CC|CXX|COMPILER` resulted in the first error again.
* Setting `CHPL_LLVM_CONFIG=llvm-config` fixed the issue.

So it seems that our scripts were auto-using `llvm-config-15` rather than the default `llvm-config` in the path for some reason.

And beyond solving that problem, things we could do to improve the debuggability of this situation were:
- [ ] improve the "Missing or wrong version for clang" error message to print more information like:
  - [ ] was llvm-config found or not?
  - [ ] if found, what was it inferred to be?
  - [ ] what were the llvm and clang versions that were considered to have mismatched?
- [ ] make `printchplenv` move past the `Missing or wrong version for clang at clang` error unless we think the previous bullet is sufficient to make progress (Maybe it is)

**Comment by bradcray** [comment]
@diehlpk : Just curious, what motivated you to start setting `CHPL_TARGET_CC|CXX` to begin with?

**Comment by bradcray** [comment]
@jabraham17 says that the answer to this part of the challenge:

> So it seems that our scripts were auto-using llvm-config-15 rather than the default llvm-config in the path for some reason.

is that our scripts prioritize `llvm-config-n` over `llvm-config-m` where `n` > `m` and `llvm-config-n` over `llvm-config` for any value of `n`.  So it seems like we could improve the situation by sorting on the _actual_ version that a given `llvm-config` represents rather than just basing the decision on the filename?  Tagging @mppf for verification that we're characterizing the logic correctly and that this idea makes sense / would be an improvement.

**Comment by jabraham17** [comment]
> in that I'd think with none of those set, we'd hopefully use the clang/clang++ associated with llvm-config-15 rather than clang/clang++ (which were version 16).

The current script logic for determining CHPL_TARGET_CC|CXX is independent from the llvm-config logic, and as best as I can tell does not consider `clang-15` when trying to infer a compiler.

**Comment by bradcray** [comment]
> Just curious, what motivated you to start setting CHPL_TARGET_CC|CXX to begin with?

@diehlpk answered this on gitter, and I thought it was worth capturing:

> Because I am used to CMake and I have to set CC and CXX to make cmake use my preferred compiler.
> So I searched in the documentation of Chapel for CC and CXX and found these.
> So I assumed I need to set these because for the first step to compile the compiler always gcc was used.
> I assumed there should be clang and not gcc.
> Since I want to compile with clang.
> Maybe it would help to mention in the documentation that gcc is always used there and clang only after that?

**Comment by bradcray** [comment]
I filed #24669 to capture what I consider the most obvious steps to be here to improve this situation for future users, and am closing this as a result.  I think we could also potentially do more to distinguish between host and target compilers/settings in the documentation, but don't have a specific proposal there at present, so am leaving that idea here for now.
